### PR TITLE
Print Attendance Roster

### PR DESCRIPTION
**Pull request by bmuniswamy**
# Critical Changes

# Changes

1. Added a new `Printable View` button on the attendance component. When the users click on the button they will be navigated to the a new tab. The new tab will display the attendance component along with session header and will have a `Print` button. When the users click that button the print preview window opens.
2. Added permissions check to the `Print` button. So the `Print` button will only be visible when the user has Read access and if there are any service delivery rows to display.
3. When the attendance records are saved the toast message will now say `Saved x Service Delivery records`
4. Added the new wire mock to jest tests

# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
- [ ] Place holder data is incorporated (Refer to [Case Management Sample Data document](https://quip.com/cbFcAPJF8t0z))
- [x] Base axe-core JEST test included for any new LWC (No critical violations)
~~- [ ] CRUD/FLS is enforced in Apex~~
~~- [ ] Permission sets & field sets are updated to account for CRUD/FLS/TABS for new object metadata~~
~~- [ ] All modified classes are unit tested (Apex) at 100% coverage~~
- [x] UX approval or UX not necessary
- [x] PR contains draft release notes
- [x] Attach work item to the pull request with [Lurch](https://salesforce.quip.com/50ZRA5LEWVzH) `**Lurch:attach W-000000` or `**Lurch:add`
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [ ] QE story level testing completed

### CODE REVIEW

**Comment by bmuniswamy** [comment]
[W-8126109](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B00000089hnUIAQ/view)
[W-8450208](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008jFNyIAM/view)

File: force-app/main/default/layouts/ServiceSession__c-Service Session Layout.layout-meta.xml [hunk 1]
```diff
@@ -2,6 +2,7 @@
 <Layout xmlns="http://soap.sforce.com/2006/04/metadata">
     <excludeButtons>ChangeOwnerOne</excludeButtons>
     <excludeButtons>ChangeRecordType</excludeButtons>
+    <excludeButtons>PrintableView</excludeButtons>
```
Comment (jjbennett) [inline force-app/main/default/layouts/ServiceSession__c-Service Session Layout.layout-meta.xml:5]: Was this change intentional or leftover from a previous iteration?

File: force-app/main/default/lwc/attendance/attendance.js [hunk 2]
```diff
@@ -31,24 +35,28 @@ import SUBMIT_LABEL from "@salesforce/label/c.Submit";
...
-import SUCCESS_MESSAGE_LABEL from "@salesforce/label/c.Save_Attendance_Success";
-import SUCCESS_NO_UPDATES_LABEL from "@salesforce/label/c.Save_Attendance_No_Updates";
+import SUCCESS_MESSAGE_LABEL from "@salesforce/label/c.Save_Service_Delivery_Success";
+import SUCCESS_NO_UPDATES_LABEL from "@salesforce/label/c.Save_Service_Delivery_No_Updates";
 import UPDATE_LABEL from "@salesforce/label/c.Update";
 import SAVE_LABEL from "@salesforce/label/c.Save";
 import CANCEL_LABEL from "@salesforce/label/c.Cancel";
 import QUANTITY_LABEL from "@salesforce/label/c.Quantity";
+import PRINT_LABEL from "@salesforce/label/c.Print";
+import PRINTABLE_VIEWABLE_LABEL from "@salesforce/label/c.Printable_View";
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendance/attendance.js:45]: Typo? Should this be PRINTABLE_VIEW_LABEL?

File: force-app/main/default/lwc/attendance/attendance.js [hunk 3]
```diff
@@ -31,24 +35,28 @@ import SUBMIT_LABEL from "@salesforce/label/c.Submit";
...
 
 const FIELD_SET_NAME = "Attendance_Service_Deliveries";
 const SHORT_DATA_TYPES = ["DOUBLE", "INTEGER", "BOOLEAN"];
 const LONG_DATA_TYPES = ["TEXTAREA", "PICKLIST", "REFERENCE"];
 const ID = "Id";
 const COMPLETE = "Complete";
 const PENDING = "Pending";
-
-export default class Attendance extends LightningElement {
+const PAGE_NAVIGATION_TYPE = "standard__navItemPage";
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendance/attendance.js:57]: Not sure this really matters, so feel free to ignore me. What about ITEM_PAGE_NAVIGATION_TYPE ? My think is that when we are reading down below we can easily identify what is stored.

File: force-app/main/default/lwc/attendanceRow/attendanceRow.js [hunk 4]
```diff
@@ -7,12 +7,11 @@
  *
  */
 
-import { LightningElement, api, track, wire } from "lwc";
+import { LightningElement, api, track } from "lwc";
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendanceRow/attendanceRow.js:10]: Nice job cleaning this up!

File: force-app/main/default/lwc/attendance/attendance.html [hunk 5]
```diff
@@ -8,6 +8,19 @@
   -->
 
 <template>
+    <lightning-card if:true={displayHeader}>
+        <div class="slds-var-m-left_medium">
+            <lightning-record-form
+                record-id={recordId}
+                object-api-name={serviceSession}
+                layout-type="Compact"
+                mode="readonly"
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendance/attendance.html:17]: Food for thought not for this iteration but I wonder if we should add breadcrumbs to make it easier to get back to the session since the Name is not clickable in this context.

**Review by jjbennett — COMMENTED** [review]
Great work, I had a couple questions. Also I am getting 2 rows in my print preview is that expected? Note:
Screen Shot 2020-11-30 at 1 57 21 PM
 I do not have a printer so it is a PDF preview.

**Comment by bmuniswamy** [comment]
@jjbennett , If we set the page layout to landscape by default you will see it in a pdf viewer. But there is an option called print user printer dialog and when you click that you can print the page. It is odd that you are seeing 2 rows since it is landscape it should have spanned the row. I will look into it

File: force-app/main/default/layouts/ServiceSession__c-Service Session Layout.layout-meta.xml (hunk 1 above)
Comment (bmuniswamy) [reply force-app/main/default/layouts/ServiceSession__c-Service Session Layout.layout-meta.xml:5]: This was intentional since we are not using the standard PrintableView button and we are using the custom one.

File: force-app/main/default/lwc/attendance/attendance.html (hunk 5 above)
Comment (bmuniswamy) [reply force-app/main/default/lwc/attendance/attendance.html:17]: Not sure if we need that since we are opening it in a new tab and they have the old tab to go back to. Let me know your thoughts?

**Review by lbeller — APPROVED** [review]

File: force-app/main/default/lwc/attendance/attendance.html (hunk 5 above)
Comment (jjbennett) [reply force-app/main/default/lwc/attendance/attendance.html:17]: Oh yea! That makes sense :)

File: force-app/main/default/layouts/ServiceSession__c-Service Session Layout.layout-meta.xml (hunk 1 above)
Comment (jjbennett) [reply force-app/main/default/layouts/ServiceSession__c-Service Session Layout.layout-meta.xml:5]: But then it looks like we have it added to the layout down below in the sorting? It is still showing on the layout.

Comment (bmuniswamy) [reply force-app/main/default/layouts/ServiceSession__c-Service Session Layout.layout-meta.xml:5]: 🤦‍♀️  I forgot to commit those changes. I am just seeing it now.

File: force-app/main/default/staticresources/pmm/attendancePrintOverride.css [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    a[href]:after {
+        content: none;
+    }
+
+    .slds-card__footer {
+        display: none;
+    }
+
+    @page {
+        size: A4 landscape;
```
Comment (jjbennett) [inline force-app/main/default/staticresources/pmm/attendancePrintOverride.css:23]: This definitely fixed my issue. Thanks!

**Review by jjbennett — APPROVED** [review]
Awesomeness, this was a tricky one for sure!
